Source/Core/Core/PowerPC/GDBThread.cpp

```diff
@@ -0,0 +1,1274 @@
+{
+  while (len-- > 0)
+  {
+    u8 tmp = HexToChar(*src++) << 4;
+    tmp |= HexToChar(*src++);
+
+    PowerPC::HostWrite_U8(tmp, dst++);
+  }
+}
+
+static void WriteHexBE32(u8* p, u32 v)
+{
+  for (u32 i = 0; i < 8; i++)
+    p[i] = NibbleToHex(u8(v >> (28 - 4 * i)));
+}
+
+static void WriteHexBE64(u8* p, u64 v)
+{
+  for (u32 i = 0; i < 16; i++)
+    p[i] = NibbleToHex(u8(v >> (60 - 4 * i)));
+}
+
+static u32 ReadHexBE32(const u8* p)
+{
+  u32 res = 0;
+
+  for (u32 i = 0; i < 8; i++)
+    res = (res << 4) | HexToChar(p[i]);
+
+  return res;
+}
+
+static u64 ReadHexBE64(const u8* p)
+{
+  u64 res = 0;
+
+  for (u32 i = 0; i < 16; i++)
+    res = (res << 4) | HexToChar(p[i]);
+
+  return res;
+}
+
+// GDB stub interface
+#ifndef _WIN32
+void GDBThread::GDBInitLocal(const char* socket)
+{
+  unlink(socket);
+
+  sockaddr_un addr = {};
+  addr.sun_family = AF_UNIX;
+  strcpy(addr.sun_path, socket);
+
+  GDBInitGeneric(PF_LOCAL, reinterpret_cast<const sockaddr*>(&addr), sizeof(addr), nullptr,
+                 nullptr);
+}
+#endif
+
+void GDBThread::GDBInit(u32 port)
+{
+  GDBDeinit();
+
+  sockaddr_in saddr_server = {};
+  sockaddr_in saddr_client;
+
+  saddr_server.sin_family = AF_INET;
+  saddr_server.sin_port = htons(port);
+  saddr_server.sin_addr.s_addr = INADDR_ANY;
+
+  socklen_t client_addrlen = sizeof(saddr_client);
+
+  GDBInitGeneric(PF_INET, reinterpret_cast<const sockaddr*>(&saddr_server), sizeof(saddr_server),
+                 reinterpret_cast<sockaddr*>(&saddr_client), &client_addrlen);
+
+  saddr_client.sin_addr.s_addr = ntohl(saddr_client.sin_addr.s_addr);
+}
+
+void GDBThread::GDBInitGeneric(int domain, const sockaddr* server_addr, socklen_t server_addrlen,
+                               sockaddr* client_addr, socklen_t* client_addrlen)
+{
+#ifdef _WIN32
+  WSADATA init_data;
+  WSAStartup(MAKEWORD(2, 2), &init_data);
+#endif
+
+  m_tmpsock = socket(domain, SOCK_STREAM, 0);
+  if (m_tmpsock == -1)
+    ERROR_LOG(GDB_THREAD, "Failed to create gdb socket");
+
+  int on = 1;
+  if (setsockopt(m_tmpsock, SOL_SOCKET, SO_REUSEADDR, reinterpret_cast<const char*>(&on),
+                 sizeof on) < 0)
+    ERROR_LOG(GDB_THREAD, "Failed to setsockopt");
+
+  if (bind(m_tmpsock, server_addr, server_addrlen) < 0)
+    ERROR_LOG(GDB_THREAD, "Failed to bind gdb socket");
+
+  if (listen(m_tmpsock, 1) < 0)
+    ERROR_LOG(GDB_THREAD, "Failed to listen to gdb socket");
+
+  NOTICE_LOG(GDB_THREAD, "Waiting for gdb to connect...\n");
+
+  m_sock = accept(m_tmpsock, client_addr, client_addrlen);
+  if (m_sock < 0)
+    ERROR_LOG(GDB_THREAD, "Failed to accept gdb client");
+  NOTICE_LOG(GDB_THREAD, "Client connected.\n");
+
+#ifdef _WIN32
+  closesocket(m_tmpsock);
+#else
+  close(m_tmpsock);
+#endif
+  m_tmpsock = -1;
+
+  m_connected = (m_sock >= 0);
+  if (!m_connected)
+    GDBDeinit();
+}
+
+void GDBThread::GDBDeinit()
+{
+  if (m_tmpsock != -1)
+  {
+    shutdown(m_tmpsock, SD_BOTH);
+#if _WIN32
```

Comment:
This if/else around `closesocket`/`close` appears fairly often, so it might be worthwhile wrapping it in a method to reduce the `#ifdef` cruft in other places (which makes it somewhat difficult to follow the intent of the code IMO; but that might just be me)

Reply:
Overall method inherited from Dolphin's GDBStub implementation, but updated to fix compile issues. Makes sense to limit duplication of the defines and logic.